File: yts.py
import requests

from collections import OrderedDict

from requests import exceptions


def build_magnet(info_hash):
    magnet = "magnet:?xt=urn:btih:"+info_hash+"&tr=udp://tracker.coppersurfer.tk:6969/announce&tr=udp://9.rarbg.to:2920/announce&tr=udp://tracker.opentrackr.org:1337&\tr=udp://tracker.internetwarriors.net:1337/announce&tr=udp://tracker.leechers-paradise.org:6969/announce&tr=udp://tracker.coppersurfer.tk:6969/announce&tr=udp://tracker.pirateparty.gr:6969/announce&tr=udp://tracker.cyberia.is:6969/announce"
    return magnet


def getJson(query):
    final_dic = OrderedDict()
    final_dic["status"] = "ok"
    final_dic["movie_count"] = 0
    final_dic["query"] = query
    final_dic["data"] = []

    url = "https://yts.am/api/v2/list_movies.json?query_term=" + \
        query+"&sort_by=download_count"
    query = query.replace(' ', '+')
    try:
        raw_data = requests.get(url).json()
    except exceptions as e:
        final_dic["error"] = str(e.message)
        return final_dic

    if raw_data["data"]["movie_count"] == 0:
        return final_dic

    final_dic["movie_count"] = raw_data["data"]["movie_count"]

    all_movies = raw_data["data"]["movies"]
    for movie in all_movies:
        try:
            name = movie["title_long"]
        except AttributeError:
            name = None
        torrents = []
        try:
            for torr in movie["torrents"]:
                try:
                    quality = torr["quality"]
                except AttributeError:
                    quality = None
                try:
                    download = torr["url"]
                except AttributeError:
                    download = None
                try:
                    hash = torr["hash"]
                    magnet = build_magnet(hash)
                except Exception:
                    hash = None
                    magnet = None
                try:
                    type = torr["type"]
                except AttributeError:
                    type = None

                try:
                    seeds = torr["seeds"]
                except AttributeError:
                    seeds = None

                try:
                    peers = torr["peers"]
                except AttributeError:
                    peers = None

                try:
                    size = torr["size"]
                except AttributeError:
                    size = None

                try:
                    date = torr["date_uploaded"]
                except AttributeError:
                    date = None

                torrObjs = {
                    "torrent_file": download,
                    "magnet": magnet,
                    "quality": quality,
                    "type": type,
                    "seeds": seeds,
                    "peers": peers,
                    "size": size,
                    "upload_date": date,
                    "hash": hash
                }
                torrents.append(torrObjs)
        except Exception:
            pass
        try:
            imdb = movie["imdb_code"]
        except AttributeError:
            imdb = None

        try:
            movie_url = movie["url"]
        except AttributeError:
            movie_url = None

        try:
            year = movie["year"]
        except AttributeError:
            year = None

        try:
            language = movie["language"]
        except AttributeError:
            language = None

        movieObj = {
            "name": name,
            "url": movie_url,
            "imdb_code": imdb,
            "year": year,
            "language": language,
            "torrents": torrents
        }

        final_dic["data"].append(movieObj)

    return (final_dic)

File: yts.js
const axios = require('axios');

function buildMagnet(infoHash) {
    return `magnet:?xt=urn:btih:${infoHash}&tr=udp://tracker.coppersurfer.tk:6969/announce&tr=udp://9.rarbg.to:2920/announce&tr=udp://tracker.opentrackr.org:1337&tr=udp://tracker.internetwarriors.net:1337/announce&tr=udp://tracker.leechers-paradise.org:6969/announce&tr=udp://tracker.pirateparty.gr:6969/announce&tr=udp://tracker.cyberia.is:6969/announce`;
}

const base64EncodeImage = async (imageUrl) => {
    try {
        const { data, headers } = await axios.get(imageUrl, { responseType: 'arraybuffer' });
        const image = Buffer.from(data, 'binary').toString('base64');
        return `data:${headers['content-type'].toLowerCase()};base64,${image}`;
    } catch (error) {
        console.error('Error encoding image:', error);
        return ''; // Return empty string on error
    }
};

async function getMovieData(query, cimage) {
    console.log("Fetching movie data for:", query);

    const finalObj = {
        status: "ok",
        movie_count: 0,
        query,
        data: []
    };

    const url = `https://yts.am/api/v2/list_movies.json?query_term=${encodeURIComponent(query)}&sort_by=download_count`;

    try {
        const { data } = await axios.get(url);
        const { movie_count, movies } = data.data;
        finalObj.movie_count = movie_count;

        if (movie_count === 0) return finalObj;

        const moviePromises = movies.map(async (movie) => {
            // Encode movie cover image
            const coverImage = cimage ? (movie.large_cover_image ? await base64EncodeImage(movie.large_cover_image) : null) : '';

            // Process torrents
            const torrents = await Promise.all(
                (movie.torrents || []).map(async (torr) => {
                    const torrentObj = {
                        torrent_file: torr.url || "",
                        magnet: torr.hash ? buildMagnet(torr.hash) : "",
                        quality: torr.quality || "",
                        type: torr.type || "",
                        seeds: torr.seeds || 0,
                        peers: torr.peers || 0,
                        size: torr.size || "",
                        upload_date: torr.date_uploaded || "",
                        hash: torr.hash || "",
                    };
                    return torrentObj;
                })
            );

            return {
                name: movie.title_long || "",
                cover_image: coverImage || "",
                description: movie.synopsis || "",
                imdb: movie.imdb_code || "",
                year: movie.year || 0,
                language: movie.language || "",
                torrents
            };
        });

        finalObj.data = await Promise.all(moviePromises);
        return finalObj;
    } catch (error) {
        finalObj.status = "error";
        finalObj.error = error.message;
        return finalObj;
    }
}

module.exports = getMovieData;
